app/controllers/spree/adyen_redirect_controller.rb

```diff
@@ -67,6 +64,19 @@ def payment_method
       @payment_method ||= Gateway::AdyenHPP.last # find(params[:merchantReturnData])
     end
 
+    def current_order
+      @__adyen_current_order ||=
+        Spree::Order.incomplete.find_by!(current_order_params)
+    end
+
+    def current_order_params
+      { number: params[:merchantReference],
+        store_id: current_store.id,
+        user_id: try_spree_current_user.try(:id),
+        currency: current_currency
```

Comment:
I'm not sure this is going to help us with the 404s, as if we come in with no state we don't know current_store, current_user and current_currency, we only know what comes in with the params. Is the order number not unique enough?